_posts/2017-11-22-macros.md

```diff
@@ -0,0 +1,291 @@
+        74: minitest.api.Asserts#intercept(Lscala/Function0;)V.
+        68: scala.async.Async.async(Lscala/Function0;Lscala/concurrent/ExecutionContext;)Lscala/concurrent/Future;.
+        66: spire.syntax.Literals#h()S.
+        65: slick.collection.heterogeneous.HList#apply(I)Ljava/lang/Object;.
+        63: org.specs2.specification.create.AutoExamples#eg(Lscala/Function0;Lorg/specs2/execute/AsResult;)Lorg/specs2/specification/core/Fragments;.
+        55: scala.StringContext#f(Lscala/collection/Seq;)Ljava/lang/String;.
+        53: org.scalatest.Matchers#StringShouldWrapper#shouldNot(Lorg/scalatest/words/CompileWord;Lorg/scalactic/source/Position;)Lorg/scalatest/compatible/Assertion;.
+        46: com.twitter.scalding.serialization.macros.LowerPriorityImplicit#primitiveOrderedBufferSupplier()Lcom/twitter/scalding/serialization/OrderedSerialization;.
+        42: spire.macros.Checked.checked(Ljava/lang/Object;)Ljava/lang/Object;.
+        37: records.Rec.fld(Lrecords/Rec;)Ljava/lang/Object;.
+        36: play.api.libs.json.Json.format()Lplay/api/libs/json/OFormat;.
+        36: utest.asserts.Asserts#compileError(Ljava/lang/String;)Lutest/CompileError;.
+        34: monix.execution.schedulers.ExecuteExtensions#executeAsync(Lscala/Function0;)V.
+        34: spire.syntax.CforSyntax#cfor(Ljava/lang/Object;Lscala/Function1;Lscala/Function1;Lscala/Function1;)V.
+        34: spire.util.Pack.longToByte(JI)B.
+        33: scodec.bits.package.BinStringSyntax#bin(Lscala/collection/Seq;)Lscodec/bits/BitVector;.
+        32: monix.execution.schedulers.ExecuteExtensions#executeTrampolined(Lscala/Function0;)V.
+        ... truncated
+```
+The complete list of results can be seen in [this gist](https://gist.github.com/olafurpg/be9d01afab4e2e54e51be620279507aa).
+Judging by the numbers, it seems that
+
+- macros are used a lot more in test sources than main sources, which is maybe expected.
+- developer ergonomics macros are heavily used in test sources, an order of magnitude more than
+  than any other category of macros.
+  This is not surprising since large test suites contain many `assert` calls.
+  Note that we excluded the ScalaTest sources from the corpus since they heavily biased
+  the results by adding 140.000 call-sites to `scalactic.source.Position.here()`.
+- optimization macros like those in spire/parboiled are widely used in main sources.
+- code generation macros like for serialization are also widely used, given that
+  the best practice is  to invoke them only once per type signature.
+
+One caveat with this analysis is that the results are heavily biased by the
+choice of the corpus, which is predominantly open source libraries.
+Please get in touch with me if you would like to add your favorite projects (esp.
+application sources!) to the corpus or learn how to run the analysis on a closed
+source project.
```

Comment:
Maybe it's best to point out here something about the results being as anonymous as possible? Well, except of course method names. ...Perhaps it's worth promising folks with closed, proprietary codebases that may be interested in sharing results with us that we won't publish their results? ...Just trying to think how we can get more info that would help our understanding of macros in the wild from closed code bases while reassuring people that we don't want to know anything about their code other than how much they use macros and for which category of use.

Reply:
Good point, I've expanded on the invitation to run the analysis on closed-source projects.